### ISSUE TITLE
Improper behaviour of 'delegate_to' option when 'with_items' with 'loop_control' and 'loop_var' are specified.

### ISSUE THREAD

**Issue by it-svit** [ansible/ansible#18961]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

delegate_to

##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.1.2.0
  config file = /home/kb/opt/infrastructure/playbooks/client/ansible.cfg
  configured module search path = Default w/o overrides

```

##### OS / ENVIRONMENT

Ansible playbook was running from the following host
```
$ cat /etc/issue
Ubuntu 16.04.1 LTS \n \l

$ uname -a
Linux skyfall 4.4.0-53-generic #74-Ubuntu SMP Fri Dec 2 15:59:10 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```
Ansible is managing the following host:
```
[~]:cat /etc/issue

Welcome to SUSE Linux Enterprise Server 12 SP1  (x86_64) - Kernel \r (\l).

[~]:uname -a
Linux 3.12.62-60.62-default #1 SMP Thu Aug 4 09:06:08 UTC 2016 (b0e5a26) x86_64 x86_64 x86_64 GNU/Linux

```
##### SUMMARY

'delegate_to' doesn't use proper vars of inventory host when 'with_items' with 'loop_control' and 'loop_var' are specified.

##### STEPS TO REPRODUCE

```
- add_host:
    hostname: "{{ item.name }}"
    groups: "sample_group"
    ansible_ssh_host: "{{ item.host }}"
    ansible_ssh_port: "{{ item.port }}"
    ansible_ssh_user: "{{ username }}"
    ansible_ssh_pass: "{{ password }}"
    ansible_become: True
    ansible_become_user: root
    ansible_become_pass: "{{ password }}"
  with_items:
  - "{{ delegated_hosts }}"

- shell: whoami; pwd
  delegate_to: "{{ delegated_item.name }}"
  with_items:
  - "{{ delegated_hosts }}"
  loop_control:
    loop_var: delegated_item
```

##### EXPECTED RESULTS

```
<host-01> ESTABLISH SSH CONNECTION FOR USER: username
<host-01> SSH: EXEC sshpass -d12 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=40001 -o User=username -o ConnectTimeout=10 -o ControlPath=/home/kb/.ansible/cp/ansible-ssh-%h-%p-%r host-01 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1481185797.44-119810802831560 `" && echo ansible-tmp-1481185797.44-119810802831560="` echo $HOME/.ansible/tmp/ansible-tmp-1481185797.44-119810802831560 `" ) && sleep 0'"'"''
<host-01> PUT /tmp/tmpY5dkzx TO /home/username/.ansible/tmp/ansible-tmp-1481185797.44-119810802831560/command
<host-01> SSH: EXEC sshpass -d12 scp -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=40001 -o User=username -o ConnectTimeout=10 -o ControlPath=/home/kb/.ansible/cp/ansible-ssh-%h-%p-%r /tmp/tmpY5dkzx '[host-01]:/home/username/.ansible/tmp/ansible-tmp-1481185797.44-119810802831560/command'
<host-01> ESTABLISH SSH CONNECTION FOR USER: username
<host-01> SSH: EXEC sshpass -d12 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=40001 -o User=username -o ConnectTimeout=10 -o ControlPath=/home/kb/.ansible/cp/ansible-ssh-%h-%p-%r host-01 '/bin/sh -c '"'"'chmod u+x /home/username/.ansible/tmp/ansible-tmp-1481185797.44-119810802831560/ /home/username/.ansible/tmp/ansible-tmp-1481185797.44-119810802831560/command && sleep 0'"'"''
<host-01> ESTABLISH SSH CONNECTION FOR USER: username
<host-01> SSH: EXEC sshpass -d12 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o Port=40001 -o User=username -o ConnectTimeout=10 -o ControlPath=/home/kb/.ansible/cp/ansible-ssh-%h-%p-%r -tt host-01 '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=zazsdpqjybdwpvclmnnfwdnoozkfxifl] password: " -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-zazsdpqjybdwpvclmnnfwdnoozkfxifl; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/username/.ansible/tmp/ansible-tmp-1481185797.44-119810802831560/command; rm -rf "/home/username/.ansible/tmp/ansible-tmp-1481185797.44-119810802831560/" > /dev/null 2>&1'"'"'"'"'"'"'"'"' && sleep 0'"'"''
changed: [localhost -> None] => (item=host-01) => {"changed": true, "cmd": "whoami; pwd", "delta": "0:00:00.005780", "end": "2016-12-08 00:30:02.533354", "invocation": {"module_args": {"_raw_params": "whoami; pwd", "_uses_shell": true, "chdir": null, "creates": null, "executable": null, "removes": null, "warn": true}, "module_name": "command"}, "item": "host-01", "rc": 0, "start": "2016-12-08 00:30:02.527574", "stderr": "", "stdout": "root\n/home/username", "stdout_lines": ["root", "/home/username"], "warnings": []}

```

##### ACTUAL RESULTS

```
<host-01> ESTABLISH SSH CONNECTION FOR USER: None
<host-01> SSH: EXEC ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/home/kb/.ansible/cp/ansible-ssh-%h-%p-%r host-01 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1481185181.57-270639976841678 `" && echo ansible-tmp-1481185181.57-270639976841678="` echo $HOME/.ansible/tmp/ansible-tmp-1481185181.57-270639976841678 `" ) && sleep 0'"'"''
failed: [localhost] (item=None) => {"delegated_item": "host-01", "msg": "Failed to connect to the host via ssh.", "unreachable": true}
```

**Comment by it-svit** [comment]
#### Workaround
```
- add_host:
    hostname: "{{ item.name }}"
    groups: "sample_group"
    ansible_ssh_host: "{{ item.host }}"
    ansible_ssh_port: "{{ item.port }}"
    ansible_ssh_user: "{{ username }}"
    ansible_ssh_pass: "{{ password }}"
    ansible_become: True
    ansible_become_user: root
    ansible_become_pass: "{{ password }}"
  with_items:
  - "{{ delegated_hosts }}"

- shell: whoami; pwd
  delegate_to: "{{ item.name }}"
  with_items:
  - "{{ delegated_hosts }}"
```

**Comment by sirkubax** [comment]
@it-svit based on the data you did provided,  I think you misunderstood the variable concept (though I may be wrong). 
Are you sure, that your "{{ delegated_hosts }}" is a list of a dictionaries?

```
delegated_hosts:
    - hostA:
        host: 'some host '
        name: 'some name'
        port: 'some port'
    - hostB:
        host: 'some host '
        name: 'some name'
        port: 'some port'

```

What is the delegated_hosts? Is this a group of hosts, or sth?
If the delegated_hosts is a group of hosts defined in the inventory, then it is just a **list of strings** like:
['hostA', 'hostB'] that would not contain the variables that you are looking for...
You would have to access your variables other way 
Example:
`hostvars["{{ item }}].name`
What's more, in your example, you are not using the **sample_group** - so not using the overrides values at all.... Please note that the overrided variables are set only in the **sample_group** (and you are using delegated_hosts)

Could you please put this at the beginig of your playbook, and paste the output??
```
  tasks:
  - debug: msg="{{ item }}"
    with_items:
    - "{{ delegated_hosts }}"

```


Anyway.....
This would be the code, that would depict fully the example that you did provide, works for me, and does not use host_vars at all (since I see no reference where you would like to load them or override...... sorry once again - in case of doubt, use `hostvars["{{ item }}].ansible_ssh_host: "{{ item.host }}"`


```
- hosts: all
  gather_facts: false

  vars:
    delegated_hosts:
      - stg-front102:
        name: 'stg-front102'
        port: 'some port'
        host: 'stg-front102'
    username: 'userxxx'
    password: 'yyy'

  tasks:
  - debug: msg="{{ item }}"
    with_items:
    - "{{ delegated_hosts }}"

  - add_host:                                                                                                                                                                                                      hostname: "{{ item.name }}"                                                                                                                                                                                         groups: "sample_group"
      ansible_ssh_host: "{{ item.host }}"
      ansible_ssh_port: "{{ item.port }}"
      #ansible_ssh_user: "{{ username }}"
      #ansible_ssh_pass: "{{ password }}"
      ansible_become: True
      ansible_become_user: root
      ansible_become_pass: "{{ password }}"
    with_items:
    - "{{ delegated_hosts }}"

  - shell: whoami; pwd
    delegate_to: "{{ delegated_item.name }}"
    with_items:
    - "{{ delegated_hosts }}"
    loop_control:
      loop_var: delegated_item

```


Please update the description, or close the bug it it is invalid :)
Have a nice day...

**Comment by sirkubax** [comment]
I can only guess, but please try (from command line)

`ansible -m ping host-01`

**Comment by it-svit** [comment]
```
$ ansible -m ping host-01
 [WARNING]: No hosts matched, nothing to do
```

No additional output.

**Comment by paulcalabro** [comment]
I believe I'm observing something related to this while using `delegate_to`, `with_items`, `loop_control`, and `loop_var`:

This works:

**File: roles/foobar/foobar.yml**
```
---
- name: Get the default IP address of all local Foobar servers
  delegate_to: "{{item}}"
  delegate_facts: true
  setup:
    filter: ansible_default_ipv4
  with_items: "{{groups['foobar-servers']}}"
  loop_control:
    loop_var: item

- name: Restrict access to the ActiveMQ transport connector to all local Foobar servers
  iptables:
    comment: "Allow traffic from {{item}}"
    chain: INPUT
    protocol: tcp
    match: tcp
    source: "{{hostvars[item]['ansible_default_ipv4']['address']}}"
    destination_port: 61617
    jump: ACCEPT
  with_items: "{{groups['foobar-servers']}}"
  loop_control:
    loop_var: item
  check_mode: true
```

But this **doesn't** work:

**File: roles/foobar/foobar.yml**
```
---
- name: Get the default IP address of all local Foobar servers
  delegate_to: "{{foobar_server}}"
  delegate_facts: true
  setup:
    filter: ansible_default_ipv4
  with_items: "{{groups['foobar-servers']}}"
  loop_control:
    loop_var: foobar_server

- name: Restrict access to the ActiveMQ transport connector to all local Foobar servers
  iptables:
    comment: "Allow traffic from {{foobar_server}}"
    chain: INPUT
    protocol: tcp
    match: tcp
    source: "{{hostvars[foobar_server]['ansible_default_ipv4']['address']}}"
    destination_port: 61617
    jump: ACCEPT
  with_items: "{{groups['foobar-servers']}}"
  loop_control:
    loop_var: foobar_server
  check_mode: true
```

I receive this error:

```
TASK [iptables : Restrict access to the ActiveMQ transport connector to all local Foobar servers] ****************************************************************************************************************
fatal: [<REDACTED>]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'dict object' has no attribute 'ansible_default_ipv4'\n\nThe error appears to have been in '<REDACTED>': line 12, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name:Restrict access to the ActiveMQ transport connector to all local Foobar servers\n  ^ here\n"}
```

...simply swapping `foobar_server` for `item` resolves the issue.

Also, here is how the playbook is being included:

File: roles/foobar/main.yml
```
---
- name: Include group playbooks
  include: "{{lookup('first_found', [group_name + '.yml', 'empty.yml'])}}"
  with_items: "{{group_names}}"
  loop_control:
    loop_var: group_name
```

Another odd thing is it looks like the setup is kinda working...

```
TASK [iptables : Get the default IP address of all local Foobar servers] *****************************************************************************************************************************************
ok: [someAMQserver.example.com] => (item=foobarServer.example.com)
```

**Comment by paulcalabro** [comment]
```
ansible 2.3.1.0
  config file = /Users/foobar/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.11 (default, Jan 22 2016, 08:29:18) [GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
```

**Comment by sivel** [comment]
This issue was resolved in bc0a0fb99ce7010d11966c805ef90ee2f0b93b99 and subsequently included in the v2.4.2 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
